File: 854-making-a-large-island/making-a-large-island.java
class Solution {
    private final int[][] DIRECTIONS = {{1, 0}, {-1, 0}, {0, 1}, {0, -1}};
    int n, m;
    int id;
    private int dfs(int[][] grid, int row, int col, int id) 
    {
        if(!isValidCell(row, col) || grid[row][col] != 1)
        return 0;
        
        // mark the cell with unique id
        grid[row][col] = id;

        // Count current land cell
        int count = 1; 
        
        // Explore all 4 directions using the DIRECTIONS array
        for(int[] dir : DIRECTIONS) 
        count += dfs(grid, row + dir[0], col + dir[1], id);
        
        return count;
    }

    public int largestIsland(int[][] grid) 
    {
        n = grid.length;
        m = grid[0].length;
        id = 2;

        // map {id, size}
        Map<Integer, Integer> map = new HashMap<>();
        map.put(0, 0); // Handle empty water cases
        
        int largestIsland = 0;
        for(int i = 0; i < n; i++) 
        {
            for(int j = 0; j < m; j++) 
            {
                if(grid[i][j] == 1) 
                {
                    int islandSize = dfs(grid, i, j, id);
                    largestIsland = Math.max(largestIsland, islandSize);
                    map.put(id, islandSize);
                    id++;
                }
            }
        }

        // flip the zero and then find the largest island
        for(int i = 0; i < n; i++) 
        {
            for(int j = 0; j < m; j++) 
            {
                if(grid[i][j] == 0) 
                {
                    // Flipping 0 to 1
                    int islandSize = 1;
                    Set<Integer> uniqueIslands = new HashSet<>();
                    for(int[] dir : DIRECTIONS) 
                    {
                        int row = i + dir[0];
                        int col = j + dir[1];

                        if(isValidCell(row, col))
                        uniqueIslands.add(grid[row][col]);
                    }

                    for(int islandId : uniqueIslands)
                    islandSize += map.get(islandId);
                    
                    largestIsland = Math.max(largestIsland, islandSize);
                }
            }
        }
        return largestIsland;
    }

    private boolean isValidCell(int row, int col)
    {
        return (row >= 0 && row < n && col >= 0 && col < m);
    }
}

File: 854-making-a-large-island/making-a-large-island.cpp
class Solution {
public:
    int n;
    vector<pair<int, int>> directions = {
        {-1, 0}, // Up
        {0, 1},  // Right
        {1, 0},  // Down
        {0, -1}  // Left
    };

    int dfs(int row, int col, vector<vector<int>>& grid, int &id)
    {
        if(row < 0 || row >= n || col < 0 || col >= n || !grid[row][col] || grid[row][col] == id)
        return 0;

        // mark the curr cell as visited
        grid[row][col] = id;

        int size = 1;
        for (auto [dx, dy] : directions) 
        {
            int r = row + dx;
            int c = col + dy;

            size += dfs(r, c, grid, id);
        }
        return size;
    }

    int largestIsland(vector<vector<int>>& grid) 
    {
        n = grid.size();
        unordered_map<int, int> map;

        int id = 2;
        // Mark each island with a unique ID and calculate its size
        for(int i = 0; i < n; i++)
        {
            for(int j = 0; j < n; j++)
            {
                if(grid[i][j] == 1)
                {
                    int islandSize = dfs(i, j, grid, id);
                    map[id] = islandSize;
                    id++;
                }
            }
        }

        // there's no land, only water exists 
        // flip any zero and get island of 1 unit
        if (map.empty()) 
        return 1;

        int maxSize = 0;
        for(int i = 0; i < n; i++)
        {
            for(int j = 0; j < n; j++)
            {
                if(grid[i][j] == 0)
                {
                    int largeIslandSize = 1;
                    unordered_set<int> set;
                    for (auto [r, c] : directions) 
                    {
                        int row = i + r;
                        int col = j + c;

                        if(row < 0 || row >= n || col < 0 || col >= n || grid[row][col] < 2)
                        continue;

                        int neighborId = grid[row][col];
                        if(!set.count(neighborId)) 
                        {
                            largeIslandSize += map[neighborId];
                            set.insert(neighborId);
                        }

                    }
                    maxSize = max(maxSize, largeIslandSize);
                }
            }
        }
        return maxSize == 0 ? n * n : maxSize;
    }
};